### PR TITLE
Modpings schedule

### LINKED ISSUE #1588: Modpings schedule feature
# Abstract

Moderators should be able to set a schedule for when their moderator ping role gets removed, to avoid waking up to a ton of unread pings.

# Specification

A new command, `!modpings schedule set [start] [stop]` (with the two arguments being UTC time), should store those two dates in a Redis cache and make sure to remove the role when `start` hits and re-add it when `stop` hits using a scheduler.

`!modpings schedule reset` should clear the cache entry and disable it.

# Reasoning

Currently, a few moderators are using `!modpings off` to not wake up to 20-50 unread mod pings and be able to focus on pings that still require their attention. This command would help by not having to remember (and sometimes forget!) to run the command every day.

### PR DESCRIPTION
**Pull request by Shivansh-007**
Closes #1588 

### Description

Screenshot from 2021-06-11 10-25-10

Screenshot from 2021-06-11 10-33-43

I have disabled the check for the minimum work time in the second test above.

### Reasoning

The PR adds another scheduler to manage the modpings schedule feature since the two features (on/off and schedule), both use the same format for the ID i.e. the moderator's discord ID. Therefore if we use the same scheduler, a moderator cannot schedule his modpings if has a task in modpings.

The modpings on/off feature would override this feature, so that if one is in `pings_off_mods` cache i.e. have there pings off for sometime, then it won't add the role on their schedule time.

### CODE REVIEW

File: bot/exts/moderation/modpings.py
Comment (Shivansh-007) [reply bot/exts/moderation/modpings.py:216]: > Instead of the user having to convert from UTC to their timezone, you can use an embed timestamp.
> 
> [Here](https://github.com/python-discord/bot/blob/main/bot/exts/moderation/modpings.py#L89-L109) is an example.

Yes, but I don't think it is possible to put two timestamps in a single embed.

Comment (Xithrius) [reply bot/exts/moderation/modpings.py:216]: This is unfortunately true.

File: bot/exts/moderation/modpings.py
Comment (Numerlor) [inline bot/exts/moderation/modpings.py:215]: Why does this use the 12 hour clock? Where time is formatted on the bot side, the iso format is used which uses the 24 hour clock

File: bot/exts/moderation/modpings.py
Comment (Numerlor) [inline bot/exts/moderation/modpings.py:84]: This method is full of errors, attach an exception logging done callback to the task to see what needs to be fixed.

**Review by Numerlor — CHANGES_REQUESTED** [review]
A quick review for now as there are some things preventing the scheduling from fully working.

The functionality should account for users leaving the moderation team and remove the scheduling when that happens.
There should also be a command to remove the schedule for yourself.

Would need confirmation as I'm not sure if I got it right myself but I don't think the 16 hour limit was meant in this way, but the other way around (as a maximum). With a 16 hour minimum the moderator will always have less than 8 hours for personal non-ping time.

File: bot/exts/moderation/modpings.py
Comment (Numerlor) [reply bot/exts/moderation/modpings.py:84]: #1637 was just merged to replace the uses of `self.bot.loop.create_task` that swallow exceptions with `scheduling.create_task(coro, event_loop=bot.loop)` to log exceptions

**Comment by Akarys42** [comment]
Hey @Shivansh-007! What is the state of this PR? Is it ready for re-review?

**Comment by Shivansh-007** [comment]
Yep it is waiting for a re-review :D

**Comment by Bluenix2** [comment]
How does this work/compare to #1570?

**Comment by Shivansh-007** [comment]
It isn't using the persistent scheduler, depending on whichever PR would get merged first, the other can do the required changes, it would be a quick change anyhow.

**Review by Bluenix2 — APPROVED** [review]

**Comment by Bluenix2** [comment]
This PR has been open for *quite a while*.. if I am gonna be honest I thought it was merged shortly after I approved it a month ago (not to be disrespectful to anyone's available time of course).

I believe this PR needs a core-dev approval (and/or another contributor approval).. let's give this some love :shipit:

File: bot/exts/moderation/modpings.py
Comment (jchristgit) [inline bot/exts/moderation/modpings.py:96]:
```suggestion
        log.trace(f"Removing moderator role from mod with ID {mod.id}")
```

**Review by jchristgit — APPROVED** [review]
This is really cool. I do think this could be even better if we were to store the modping schedules in the PostgreSQL database of the site, because then we can query it and inspect there, but that would just be a "nice to have".

**Review by ChrisLovering — APPROVED** [review]
Proxy +1 for Volcyy

**Pull request merged**
